include/zephyr/devicetree.h

```diff
@@ -2159,14 +2159,29 @@
 #define DT_REG_HAS_IDX(node_id, idx) \
 	IS_ENABLED(DT_CAT4(node_id, _REG_IDX_, idx, _EXISTS))
 
+/**
+ * @brief Get the base address of the register block at index @p idx
+ *		in a raw form, without integer literal suffix like LLU
+ * @param node_id node identifier
+ * @param idx index of the register whose address to return
+ * @return address of the idx-th register block in a raw form
+ */
+#define DT_REG_ADDR_BY_IDX_RAW(node_id, idx) \
+	DT_CAT4(node_id, _REG_IDX_, idx, _VAL_ADDRESS)
+
 /**
  * @brief Get the base address of the register block at index @p idx
  * @param node_id node identifier
  * @param idx index of the register whose address to return
  * @return address of the idx-th register block
  */
+#if IS_ENABLED(CONFIG_64BIT)
 #define DT_REG_ADDR_BY_IDX(node_id, idx) \
-	DT_CAT4(node_id, _REG_IDX_, idx, _VAL_ADDRESS)
+	UINT64_C(DT_REG_ADDR_BY_IDX_RAW(node_id, idx))
+#else
+#define DT_REG_ADDR_BY_IDX(node_id, idx) \
+	DT_REG_ADDR_BY_IDX_RAW(node_id, idx)
+#endif /* IS_ENABLED(CONFIG_64BIT) */
```

Comment:
I'm not convinced about this solution. `DT_REG_ADDR*` is also used for non-64-bit buses in 64-bit systems, e.g. I2C or SPI for CS. I think we should have _typed_ macros instead, e.g. `DT_REG_ADDR_U64`. For props, `dt-schema` has fixed-integer types: https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/types.yaml we should likely adopt them and offer typed versions of the DT API macros, e.g. `DT_PROP_U32`, etc.

Reply:
So guys, what will we decide? To return to boilerplate code using DT_REG_ADDR_U64 which will be a wrapper around UINT64_C macro, right? And put it to all the places where necessary?

E.g. like this:

```
#if IS_ENABLED(CONFIG_64BIT)
#define MTIMEDIV_REG	(DT_INST_REG_ADDR_U64(0)) + 4)
#define MTIME_REG	(DT_INST_REG_ADDR_U64(0) + 8)
#define MTIMECMP_REG	(DT_INST_REG_ADDR_U64(0) + 16)
#else
#define MTIMEDIV_REG	(DT_INST_REG_ADDR(0) + 4)
#define MTIME_REG	(DT_INST_REG_ADDR(0) + 8)
#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 16)
#endif /* CONFIG_64BIT */
```

@gmarull , @carlocaione , @evgeniy-paltsev , @mbolivar-nordic

Reply:
not ideal but this looks reasonable to me.

Reply:
> Just for note - I'd say that `CONFIG_64BIT` is not related to peripheral addresses.
> 
> We may have 32bit peripheral bus (or peripheral which require 32bit / 16bit / 8bit access) on 64bit CPU and we can have 64bit peripheral bus with 32bit CPU.

Agree. Moreover, adding ULL to the address constant which is in fact limited to 32-/16-/8-bit seems to be safe, so basically we can use _U64 version unconditionally everywhere where sometimes the real 64-bit address may be passed from device tree.  And the compiler will complain in case of overflow, as before. 

So basically returning to my example above we may just stick to:
```
#define MTIMEDIV_REG	(DT_INST_REG_ADDR_U64(0) + 4)
#define MTIME_REG	(DT_INST_REG_ADDR_U64(0) + 8)
#define MTIMECMP_REG	(DT_INST_REG_ADDR_U64(0) + 16)
```

Right?

Reply:
>  I think we should have typed macros instead, e.g. DT_REG_ADDR_U64

This seems reasonable to me -- @galak?